File: Backend/preprocessLogs.py
import sys
import pandas as pd
import re

# ✅ FINAL COLUMN ORDER
OUTPUT_COLUMNS = [
    "Date and Time",
    "Event ID",
    "Level",
    "Source",
    "UtcTime",
    "Process Name",
    "Process Path",
    "Process ID",
    "Parent Process Name",
    "Command Line",
    "User / Integrity Level",
    "Hashes",
]

def find_raw_column(df):
    for col in df.columns:
        if "Unnamed" in str(col):
            return col

    for col in df.columns:
        sample = df[col].astype(str).head(20)
        if sample.str.contains(r"[A-Za-z]+: ").sum() > 10:
            return col

    raise Exception("Raw column not found")

def parse_body(raw):
    if not isinstance(raw, str):
        return {}

    raw = raw.replace("\r\n", "\n").replace("\r", "\n")

    result = {}
    lines = raw.split("\n")

    current_key = ""
    current_value = []

    for line in lines:
        match = re.match(r"^([A-Za-z0-9]+):\s*(.*)", line)

        if match:
            if current_key:
                result[current_key] = " ".join(current_value).strip()

            current_key = match.group(1)
            current_value = [match.group(2)]
        else:
            current_value.append(line.strip())

    if current_key:
        result[current_key] = " ".join(current_value).strip()

    return result

def preprocess(input_path, output_path):
    print("🔥 PYTHON SCRIPT STARTED")

    df = pd.read_csv(input_path, dtype=str, low_memory=False)

    # 🔥 FIX COLUMN SHIFTING (CRITICAL)
    df["Real_Date"] = df["Level"]
    df["Real_Source"] = df["Date and Time"]
    df["Real_EventID"] = df["Source"]

    # 🔥 RAW BODY COLUMN
    raw_col = "Task Category" if "Task Category" in df.columns else find_raw_column(df)

    records = []

    for _, row in df.iterrows():
        record = {}

        body = parse_body(row.get(raw_col, ""))

        # ✅ FINAL STRUCTURE (ORDER MATTERS)
        record["Date and Time"] = row.get("Real_Date", "")
        record["Event ID"] = row.get("Real_EventID", "")
        record["Level"] = "Information"  # constant in your logs
        record["Source"] = row.get("Real_Source", "")

        record["UtcTime"] = body.get("UtcTime", "")

        image = body.get("Image", "")
        record["Process Name"] = image
        record["Process Path"] = image

        record["Process ID"] = body.get("ProcessId", "")
        record["Parent Process Name"] = body.get("ParentImage", "")
        record["Command Line"] = body.get("CommandLine", "")

        user = body.get("User", "")
        integrity = body.get("IntegrityLevel", "")
        # record["User / Integrity Level"] = f"{user} | {integrity}".strip(" |")
        record["User / Integrity Level"] = f"{user} | {integrity}".strip(" |")

        record["Hashes"] = body.get("Hashes", "")

        records.append(record)

    df_out = pd.DataFrame(records, columns=OUTPUT_COLUMNS).fillna("")

    df_out.to_csv(output_path, index=False)

    print("✅ PREPROCESSING DONE")

if __name__ == "__main__":
    input_file = sys.argv[1]
    output_file = sys.argv[2]
    preprocess(input_file, output_file)

File: Backend/src/utils/preprocessLogs.ts
import { parse } from "csv-parse/sync";

/* -----------------------------
   CONFIG
----------------------------- */

const META_COLS = ["Level", "Date and Time", "Source", "Event ID"];

const BODY_FIELD_MAP: Record<string, string> = {
  UtcTime: "UtcTime",
  Image: "Process Name (Image)",
  ProcessId: "Process ID",
  CommandLine: "Command Line",
  ParentImage: "Parent Process Name",
  Hashes: "Hashes",
};

const OUTPUT_COLUMNS = [
  "Level",
  "Date and Time",
  "Source",
  "Event ID",
  "UtcTime",
  "Process Name (Image)",
  "Process ID",
  "Command Line",
  "Parent Process Name",
  "Hashes",
];

/* -----------------------------
   HELPERS
----------------------------- */

// 🔍 Find raw body column
function findRawBodyColumn(rows: any[]): string {
  if (!rows.length) throw new Error("Empty CSV");

  const columns = Object.keys(rows[0]);

  // Strategy 1 → Unnamed column
  for (const col of columns) {
    if (/^Unnamed/i.test(col)) {
      console.log("Raw column (Unnamed):", col);
      return col;
    }
  }

  // Strategy 2 → heuristic (contains key:value)
  for (const col of columns) {
    if (META_COLS.includes(col)) continue;

    const sample = rows.slice(0, 20).map(r => String(r[col] || ""));

    const count = sample.reduce((acc, val) => {
      return acc + ((val.match(/[A-Za-z]+:\s/g) || []).length);
    }, 0);

    if (count > 10) {
      console.log("Raw column (heuristic):", col);
      return col;
    }
  }

  throw new Error("Raw Sysmon body column not found");
}

// 🧠 Parse raw body into key-value
function parseBody(raw: string): Record<string, string> {
  if (!raw || typeof raw !== "string") return {};

  const normalized = raw
    .replace(/\r\n/g, "\n")
    .replace(/\r/g, "\n");

  const parts = normalized.split(/\n([A-Za-z][A-Za-z0-9]*):\s*/);

  const result: Record<string, string> = {};

  for (let i = 1; i < parts.length; i += 2) {
    const key = parts[i]?.trim();
    const value = parts[i + 1]?.trim();

    if (key) result[key] = value;
  }

  return result;
}

// 🧩 Build final structured record
function buildRecord(row: any, rawCol: string) {
  const record: any = {};

  // Meta fields
  for (const col of META_COLS) {
    record[col] = row[col] || "";
  }

  // Body fields
  const body = parseBody(row[rawCol] || "");

  for (const [rawKey, outKey] of Object.entries(BODY_FIELD_MAP)) {
    record[outKey] = body[rawKey] || "";
  }

  return record;
}

/* -----------------------------
   MAIN FUNCTION
----------------------------- */

export function preprocessRawSysmon(content: string) {
  // Step 1: Parse CSV
  const rows = parse(content, {
    columns: true,
    skip_empty_lines: true,
    relax_column_count: true,
    relax_quotes: true,
  });

  if (!rows.length) return [];

  // Step 2: Find raw column
  const rawCol = findRawBodyColumn(rows);

  // Step 3: Convert each row
  const processed = rows.map((row: any) =>
    buildRecord(row, rawCol)
  );

  // Step 4: Ensure column consistency
  return processed.map((row: any) => {
    const clean: any = {};
    for (const col of OUTPUT_COLUMNS) {
      clean[col] = row[col] || "";
    }
    return clean;
  });
}